File: SortingAlgorithmsImpl/QuickSort.java
public class QuickSort {

	QuickSort() {}

	
	public void sort(int arr[], int low, int high) {

		if (low < high) {

			int pivot = partition(arr, low, high);
				
			sort(arr, low, pivot-1);
			sort(arr, pivot+1, high);
		}

	}

	public int partition(int arr[], int low, int high) {

		int pivot = arr[high];

		int i = low;

		for (int j = low; j < high; j++) {

			if (arr[j] <= pivot) {
				int temp = arr[j];
				arr[j] = arr[i];
				arr[i] = temp;
				i++;				
			}		
		}
		int temp = arr[high];
		arr[high] = arr[i];
		arr[i] = temp;
		pivot = i;
		return pivot;
		
	}

	public static void main(String[] args) {

		int arr[] = {5, 2, 10, 7, 4};
		int n = arr.length;

		for (int ele: arr) {
			System.out.print(ele + " ");
		}

		System.out.println();

		QuickSort quickSort = new QuickSort();
		quickSort.sort(arr, 0, n-1);

		for (int ele: arr) {
			System.out.print(ele + " ");
		}		
	}
}

File: SortingAlgorithmsImpl/QuickSort.cpp
#include <iostream>
#include <vector> 
#include "/Algos/HelperClass/Vecs.h"

using namespace std;

class QuickSort {

	public:

		static void quickSort(vector<int> &vec, int low, int high) {

			if (low >= high) {
				return;
			}
			size_t pivot = partition(vec, low, high); 

			quickSort(vec, low, pivot - 1);
			quickSort(vec, pivot + 1, high);
		}
		
		static int partition(vector<int> &vec, int low, int high) {

			int pivot = vec[high];
			int i = low; 
			
			for (size_t j = (size_t)low; j < (size_t)high; j++) {

				if (vec[j] <= pivot) {
					int temp = vec[i];
					vec[i] = vec[j];
					vec[j] = temp;
					i++;
				}
			}

			int temp = vec[high];
			vec[high] = vec[i];
			vec[i] = temp;
			pivot = i;

			return pivot; 
		}	
};

int main() {

	Vecs vecs;
	QuickSort qs;

	vector<int> vec = {3, 10, 4, 1, 8};
	int low = 0;
	int high = 4;

	vecs.printVector(vec);
	
	qs.quickSort(vec, low, high);

	vecs.printVector(vec);
}
